### PR TITLE
[SPARK-40320][Core] Executor should exit when initialization failed for fatal error

### PR DESCRIPTION
**Pull request by yabola**
### What changes were proposed in this pull request?
Before when the Executor plugin fails to initialize for fatal error, the Executor shows active in sparkUI but does not accept any job. As expected Executor should exit when initialization failed with fatal error. 
The root cause:  The thread pool execution uses the inappropriate API `submit` instead of `execute`. This leads to  `SparkUncaughtExceptionHandler` can't catch the fatal error.

### Why are the changes needed?
As expected Executor should exit when initialization failed with fatal error. Before Executor was active but can't do anything which is confused.

### Does this PR introduce any user-facing change?
No
### How was this patch tested?
new added test

### CODE REVIEW

**Comment by yabola** [comment]
@HyukjinKwon Hi, if you hive time, can you help take a look and see which way is suitable

**Comment by yabola** [comment]
To answer why `uncaughtExceptionHandler` in `Executor` doesn't catch error 
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L106 

I have a simple little program to explain

`
object TestExceptionHandler {

  val threadFactory = new ThreadFactoryBuilder().setDaemon(true).build()

  def main(args: Array[String]): Unit = {
    val service = Executors.newSingleThreadExecutor(threadFactory)
    val runLoop = new Runnable() {
      override def run(): Unit = receiveLoop()
    }
    service.execute(runLoop)
    Thread.sleep(2 * 1000)
  }

  private def receiveLoop() {
    try {
      Executors.newSingleThreadExecutor(threadFactory).execute(new MessageLoop)
    } catch {
      case t: Throwable => {
        println("`receiveLoop` catch the Exception")
      }
    }
  }


  class MessageLoop extends Runnable {
    Thread.setDefaultUncaughtExceptionHandler(new MyExceptionHandler)
    /**
     * This Exception will not be caught by `MyExceptionHandler`. It will be caught by `receiveLoop`
     */
    if (1 == 1) {
      throw new UnsatisfiedLinkError("My Exception error, will be caught by receiveLoop")
    }
    override def run(): Unit = {
      println("MessageLoop run method.... ")
      /**
       * This Exception will be caught by `MyExceptionHandler`
       * (The premise is that the above Exception code needs to be commented out)
       */
      if (1 == 1) {
        throw new UnsatisfiedLinkError("My Exception error, will be caught by `MyExceptionHandler` !")
      }
    }
  }

  class MyExceptionHandler extends Thread.UncaughtExceptionHandler {
    override def uncaughtException(t: Thread, e: Throwable): Unit = {
      println("`MyExceptionHandler` catch the Exception... ")
    }
  }
}
`

**Comment by yabola** [comment]
object TestExceptionHandler {

  val threadFactory = new ThreadFactoryBuilder().setDaemon(true).build()

  def main(args: Array[String]): Unit = {
    val service = Executors.newSingleThreadExecutor(threadFactory)
    val runLoop = new Runnable() {
      override def run(): Unit = receiveLoop()
    }
    service.execute(runLoop)
    Thread.sleep(2 * 1000)
  }

  private def receiveLoop() {
    try {
      Executors.newSingleThreadExecutor(threadFactory).execute(new MessageLoop)
    } catch {
      case t: Throwable => {
        println("receiveLoop catch the Exception")
      }
    }
  }


  class MessageLoop extends Runnable {
    Thread.setDefaultUncaughtExceptionHandler(new MyExceptionHandler)
    /**
     * This Exception will not be caught by MyExceptionHandler. It will be caught by receiveLoop
     */
    if (1 == 1) {
      throw new UnsatisfiedLinkError("My Exception error, will be caught by receiveLoop")
    }
    override def run(): Unit = {
      println("MessageLoop run method.... ")
      /**
       * This Exception will be caught by MyExceptionHandler
       * (The premise is that the above Exception code needs to be commented out)
       */
      if (1 == 1) {
        throw new UnsatisfiedLinkError("My Exception error, will be caught by MyExceptionHandler !")
      }
    }
  }

  class MyExceptionHandler extends Thread.UncaughtExceptionHandler {
    override def uncaughtException(t: Thread, e: Throwable): Unit = {
      println("MyExceptionHandler catch the Exception... ")
    }
  }
}

**Comment by yabola** [comment]
Then say the above it did catch the fatal error in https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala#L82-L89 

It will resubmit receiveLoop and after that it will be blocked here https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala#L69

This Executor did not initialize successfully in the first time , so it didn't send LaunchedExecutor to Driver (you can see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L172 )

So the Executor can't launch task and can't do anything now (even worker is in active state in spark UI). It is very hard to know what happened here unless we check in the code why it hung here.

Actually before the Executor will exit when launch task if Executor can't initialize . However now if Executor can't initialize then the worker won't launch any task so the Executor won't exit now. (you can see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L180
and PR : https://github.com/apache/spark/pull/25964)

**Comment by yabola** [comment]
@Ngone51 Hi, I updated some reasons, I don't know if it can be notified on jira, so I updated it in PR

**Comment by yabola** [comment]
@mridulm @tgravescs If you have time, can you also help take a look~

**Comment by mridulm** [comment]
I am not sure I follow what the code snippet is trying to do.
Changing the code to :

```
  private def receiveLoop() {
      Executors.newSingleThreadExecutor(threadFactory).execute(new MessageLoop)
    }
```
does cause the `uncaughtException` to be invoked - as expected.
If we catch the `Throwable` in `receiveLoop` before that, then obviously it is no longer an uncaught exception - and so `MyExceptionHandler.uncaughtException` wont be invoked.

**Comment by yabola** [comment]
@mridulm 
Yes, It is another situation, there are three cases to explain whether uncaughtException can catch exceptions.

1. 
As you said
```
private def receiveLoop() {
      Executors.newSingleThreadExecutor(threadFactory).execute(new MessageLoop)
    }
```
uncaughtException will be invoked.

2. 
If we surrounded with try catch , then uncaughtException won't be invoked.
```
private def receiveLoop() {
    try {
      Executors.newSingleThreadExecutor(threadFactory).execute(new MessageLoop)
    } catch {
      case t: Throwable => {
        println("receiveLoop catch the Exception")
      }
    }
  }
```
3.
Even If we surrounded with try-catch , uncaughtException can be invoked if Exception happened in `TestExceptionHandler.MessageLoop#run()`
But we need to remove the first Exception code in `MessageLoop`( otherwise it could not be initialized ).

So it can also explain the behavior in Spark. 
`org.apache.spark.rpc.netty.MessageLoop#receiveLoop` is surrounded with try-catch, so when error happened in Executor init , uncaughtException can't catch the fatal error.
But if Executor initialize successfully, uncaughtException can catch error in `org.apache.spark.executor.Executor.TaskRunner#run`

**Comment by yabola** [comment]
My example answer the question from @Ngone51 in Jira.  But back to the current issue of this,

> It did catch the fatal error in https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala#L82-L89
> 
> It will resubmit receiveLoop and after that it will be blocked here https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala#L69
> 
> This Executor did not initialize successfully in the first time , so it didn't send LaunchedExecutor to Driver (you can see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L172 )
> 
> So the Executor can't launch task and can't do anything now (even Executor is in active state in sparkUI). It is very hard to know what happened here unless we check in the code why it hung here.
> 
> Actually before the Executor will exit when launch task if Executor can't initialize . However now if Executor can't initialize then the worker won't launch any task so the Executor won't exit now. (you can see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L180 and PR : #25964)

I think we would better catch the fatal error when initialize Executor as my PR changed

**Comment by Ngone51** [comment]
https://github.com/apache/spark/blob/39b65b414c4ba36ada478369149f54452d90dd7b/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L169-L176


The issue seems to be that `Executor` construction failed due to the fatal error thrown during plugin initialization. And the fatal error doesn't fail the executor process, which leaves the executor in idle and becomes unavailable to the driver since `LaunchedExecutor` is not sent. Correct me if I'm wrong.

**Comment by mridulm** [comment]
A few points:

* Plugins are initialized as part of `Executor` construction, but after uncaught exception handler is set.
* If anything other than `NonFatal` is thrown in `receive` (like the `LinkageError` from @yabola's example here), it will be handled in `Inbox.process` as part of `safelyCall` - which will result in re-throwing it.
* This now bubbles up to `receiveLoop` - where we end up throwing the throwable again (after scheduling it for re-execution).
* The throw should result in uncaught exception handler killing the jvm - and if it does not, then the re-enqueue in prev step will cause the message to be reprocessed.
  * In this case we are discussing, it should have resulted in killing the VM.

**Comment by yabola** [comment]
@Ngone51 @mridulm Yeah~You are right. Do you have any idea to solve this problem? My idea is to just exit Executor as PR changed.

**Comment by mridulm** [comment]
I will try to reproduce this locally - is there anything else specific I need to know to do so ? Or is the instructions in SPARK-40320 sufficient ? Thx

**Comment by yabola** [comment]
@mridulm I think it is sufficient. If you have any doubts please tell me~
I run my test code in spark standalone cluster(It should also have this problem with other modes)
This is my driver code:
```
val spark = SparkSession
      .builder()
      .appName("plugin example")
      .config("spark.plugins", "org.apache.spark.examples.sql.ErrorSparkPlugin")
      .getOrCreate()
    val df = spark.range(10, 20)
    df.show()
    Thread.sleep(20 * 60 * 1000)
    spark.stop()
```

**Comment by Ngone51** [comment]
> The throw should result in uncaught exception handler killing the jvm - and if it does not, then the re-enqueue in prev step will cause the message to be reprocessed.

@mridulm Is the fatal error caught by the uncaught exception handler in this case? If yes, why it doesn't fail the JVM?

**Comment by mridulm** [comment]
I agree @Ngone51, it should have :-) I am trying to reproduce this locally and see if we are missing something - or there is some nuance behind the observed behavior.

**Comment by mridulm** [comment]
Added a few debug statements, and it became clear what the issue is.
Essentially, since we are leveraging a `ThreadPoolExecutor`, it does not result in killing the thread with the exception/error thrown - but rather, will call `ThreadPoolExecutor.afterExecute` with the cause for failure (See `runWorker` for more).

We should be overriding this, and invoke our `uncaughtExceptionHandler` when an exception is thrown.


In `receiveLoopRunnable` when a `Throwable` is thrown:

```
22/09/26 17:17:12 INFO DedicatedMessageLoop: Current exceptionHandler = org.apache.spark.util.SparkUncaughtExceptionHandler@27c71f14
22/09/26 17:17:12 INFO DedicatedMessageLoop: Thread = Thread[dispatcher-Executor,5,main]
22/09/26 17:17:12 INFO DedicatedMessageLoop: Stack ...
java.lang.Exception: For stack
        at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:56)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

```

In `receiveLoopRunnable`'s `run`, when a `Throwable` is thrown:
```
2/09/26 17:17:12 INFO DedicatedMessageLoop: Thread = Thread[dispatcher-Executor,5,main], stackTrace =
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.lang.Thread.dumpThreads(Native Method)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.lang.Thread.getAllStackTraces(Thread.java:1653)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     app//org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$dumpAllStackTraces(MessageLoop.scala:70)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     app//org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:58)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.FutureTask.run(FutureTask.java:264)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
22/09/26 17:17:12 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)

```


and finally, a few seconds after Executor Inbox failure - dumping all threads in a new thread.
```
22/09/26 17:17:14 INFO DedicatedMessageLoop: Thread = Thread[dispatcher-Executor,5,main], stackTrace =
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/jdk.internal.misc.Unsafe.park(Native Method)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2081)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:433)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     app//org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:102)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     app//org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:45)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
22/09/26 17:17:14 INFO DedicatedMessageLoop:     java.base@11.0.16/java.lang.Thread.run(Thread.java:829)
```

**Comment by Ngone51** [comment]
> Essentially, since we are leveraging a ThreadPoolExecutor, it does not result in killing the thread with the exception/error thrown - but rather, will call ThreadPoolExecutor.afterExecute with the cause for failure (See runWorker for more).


@mridulm  Thanks for your analysis.  But I still don't get why `SparkUncaughtExceptionHandler` doesn't catch the throwable. 

```java
try {
  beforeExecute(wt, task);
  Throwable thrown = null;
  try {
      task.run();
  } catch (RuntimeException x) {
      thrown = x; throw x;
  } catch (Error x) {
      thrown = x; throw x;
  } catch (Throwable x) {
      thrown = x; throw new Error(x);
  } finally {
      afterExecute(task, thrown);
  }
} finally {
  task = null;
  w.completedTasks++;
  w.unlock();
}
```
`afterExecute` does handle the throwable but it doesn't catch it. So the throwable should be still thrown out.

**Comment by mridulm** [comment]
Thanks for the query @Ngone51 - I missed out one aspect of my analysis, which ends up completely changing the solution - my bad :-(

The answer to your query has the reason for the lack of failure - this is due to the two types of api's we are using ...

For `DeducatedMessageLoop`, for the initial submission of `receiveLoopRunnable` - we use `ExecutorService.submit` api - while for other cases, we use the `Executor.execute` api - and this is the cause for the behavior.

The `submit` api returns a `Future` - and if we look at the implementation of `FutureTask.run`, we see that it catches `Throwable` and preserves that as the outcome (`setException`) - which is the reason why the thread itself does not die.


So the specific case might be mitigated by using `execute` instead of `submit` api.

**Comment by Ngone51** [comment]
Oh I see. That's really a subtle diff.  Thanks @mridulm 

> So the specific case might be mitigated by using execute instead of submit API.

+1

**Comment by mridulm** [comment]
Can you take a look at comment above @yabola and work on the fix ? Since you already spent a lot of time on this.

**Comment by yabola** [comment]
@mridulm Really thanks for your analysis! Please give me some time to understand.

**Comment by yabola** [comment]
@mridulm Thank you very much for your detailed analysis. I had tested it in my cluster and it worked as expected.

**Comment by mridulm** [comment]
Can we also add the example code you had to reproduce the issue as a test ?

File: core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala [hunk 1]
```diff
@@ -171,7 +171,7 @@ private class DedicatedMessageLoop(
   }
 
   (1 to endpoint.threadCount()).foreach { _ =>
-    threadpool.submit(receiveLoopRunnable)
+    threadpool.execute(receiveLoopRunnable)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala:174]: Could you add comments to explain why we change to `execute`?

**Comment by yabola** [comment]
@mridulm The test class needs to run in the local-cluster mode to simulate, see if it is suitable, thanks~

File: core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala [hunk 2]
```diff
@@ -535,6 +538,40 @@ class CoarseGrainedExecutorBackendSuite extends SparkFunSuite
...
+        executorAddCounter.getAndIncrement()
+      }
+
+      override def onExecutorRemoved(executorRemoved: SparkListenerExecutorRemoved): Unit = {
+        executorRemovedCounter.getAndIncrement()
+      }
+    }
+    try {
+      sc.addSparkListener(listener)
+      eventually(timeout(60.seconds)) {
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala:566]: Do we want to run this for 60s ? Perhaps, reduce it to 10s ?
Thoughts @Ngone51 ?

**Review by mridulm — APPROVED** [review]
Looks good to me, thanks for working on this @yabola !
Just a couple of minor comments.

+CC @Ngone51

File: core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala [hunk 3]
```diff
@@ -171,7 +171,11 @@ private class DedicatedMessageLoop(
   }
 
   (1 to endpoint.threadCount()).foreach { _ =>
-    threadpool.submit(receiveLoopRunnable)
+    /**
+     * We need to be careful not to use [[ExecutorService#submit]].
+     * `submit` api will swallow errors in [[FutureTask#setException]].
```
[the remark this answers is not in the document]
Comment (yabola) [reply core/src/main/scala/org/apache/spark/rpc/netty/MessageLoop.scala]: done~

File: core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala (hunk 2 above)
Comment (yabola) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: This test takes about 5s. I make the timeout time longer here to hope that the UT will be more stable. When the machine load is heavy, the time-consuming time may become longer. 
I don't know if I need to worry about these, how about the modification to 20s.

**Comment by yabola** [comment]
@Ngone51 Hi~If you have time, please have a look at this change, thanks

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
+CC @Ngone51
I will merge it later tomorrow if no further comments.

**Comment by mridulm** [comment]
Looks like this dropped off my radar - merging to master.

**Comment by mridulm** [comment]
Thanks for working on this @yabola !
Thanks for the reviews and brainstorming on this @Ngone51 :-)

**Comment by mridulm** [comment]
I am assuming your jira id is `miracle` based on previous PR conversation - do let me know if that is incorrect @yabola !

**Comment by dongjoon-hyun** [comment]
Thank you for this fix, all!

**Comment by yabola** [comment]
@mridulm @Ngone51 Thank you all for reviews!
Yeah, I think my jira id is `miracle`. I checked my jira:  Login name and `Username` is `miracle`, but `Full name` is `Mars`.

**Pull request closed without merging**